File: backend/app/schemas/news.py
"""Схемы для вкладки «Новинки» / афиша в кино."""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import ImageURLs


class NewsCinemaHint(BaseModel):
    """Кинотеатр, где идёт фильм (из парсинга афиши)."""

    name: str


class NewsFilmItem(BaseModel):
    """Фильм в прокате в городе пользователя."""

    kinopoisk_id: int | None = None
    title: str
    entity_id: int | None = None
    release_year: int | None = None
    summary: str | None = None
    images: ImageURLs = ImageURLs()
    ticket_url: str
    ticket_provider: Literal["kinopoisk"] = "kinopoisk"
    cinemas: list[str] = Field(default_factory=list)
    in_database: bool = False
    tmdb_id: int | None = None

    model_config = ConfigDict(from_attributes=True)


class NewsResponse(BaseModel):
    city: str
    city_kp_id: str
    source: Literal[
        "kinopoisk_afisha", "tmdb_fallback", "tmdb_world", "tmdb_upcoming"
    ] = "kinopoisk_afisha"
    fetched_at: datetime
    items: list[NewsFilmItem]
    total: int
    limit: int

File: front/src/api/news.ts
import { apiClient } from './client';
import type { ImageURLs } from './types';

export interface NewsCity {
  name: string;
  slug: string;
  kp_city_id: string;
}

export interface NewsFilmItem {
  kinopoisk_id: number | null;
  title: string;
  entity_id: number | null;
  release_year: number | null;
  summary: string | null;
  images: ImageURLs;
  ticket_url: string;
  ticket_provider: 'kinopoisk';
  cinemas: string[];
  in_database: boolean;
  tmdb_id: number | null;
}

export interface NewsResponse {
  city: string;
  city_kp_id: string;
  source: string;
  fetched_at: string;
  items: NewsFilmItem[];
  total: number;
  limit: number;
}

export interface ListNewsParams {
  scope?: 'city' | 'world';
  city?: string;
  lang?: 'ru' | 'en';
  limit?: number;
}

export async function listNews(params: ListNewsParams = {}): Promise<NewsResponse> {
  const { data } = await apiClient.get<NewsResponse>('/api/news', { params });
  return data;
}

export async function listUpcoming(limit = 8): Promise<NewsResponse> {
  const { data } = await apiClient.get<NewsResponse>('/api/news/upcoming', {
    params: { limit },
  });
  return data;
}

export async function getFilmAfisha(
  entityId: number,
  city?: string,
): Promise<NewsFilmItem> {
  const { data } = await apiClient.get<NewsFilmItem>(`/api/news/film/${entityId}`, {
    params: city ? { city } : undefined,
  });
  return data;
}

export async function listNewsCities(): Promise<NewsCity[]> {
  const { data } = await apiClient.get<NewsCity[]>('/api/news/cities');
  return data;
}
